dbt/include/spark/macros/adapters.sql

```diff
@@ -5,13 +5,36 @@
     {{ sql }}
 {% endmacro %}
 
+
 {% macro file_format_clause() %}
   {%- set file_format = config.get('file_format', validator=validation.any[basestring]) -%}
   {%- if file_format is not none %}
     using {{ file_format }}
   {%- endif %}
 {%- endmacro -%}
 
+
+{% macro location_clause() %}
+  {%- set path = config.get('location', validator=validation.any[basestring]) -%}
+  {%- if path is not none %}
+    location '{{ path }}'
```

Comment:
Question on this.... If I understand this correctly, the location is an absolute reference to the specific table, yes?

Do you have any thoughts to a use case where we would declare a root_location which would be suffixed with the table name? I can see the merits of being able to override the location on a single table but I think it would also be very helpful for the entire project (or schema) to have a common `base_location` or `location_root`.

Example usage using `location_root`:

```
location_root = s3://my-bucket/data/out/
```

Given a table name 'my_table', the storage location would be: `s3://my-bucket/data/out/my_table`.

Alternatively, and I don't know if this is possible/easy with jinja templates, can we parse the table name into the location in a parameterized manner?

Example:
```
location = s3://my-bucket/data/out/{{ table_name }}
```

...which (I think) would provide the same output as the previous example.

Can you confirm if this is use case is supported and if so, what the expected behavior/usage would be?

Thanks!

Reply:
Hi Aaron, that would be a great feature. Unfortunately, I don't know how to actually implement it. I guess adding a location root could work, but I guess will also need to add the data as schema to avoid conflicts

Reply:
@NielsZeilemaker - I think this might come "for free" but I'm trying to confirm. 

See related thread I've created in slack: https://getdbt.slack.com/archives/C50NEBJGG/p1580240287016800

Reply:
Per our conversation in Slack, it's possible to dynamically include the model name with the special Jinja variable [`this`](https://docs.getdbt.com/docs/this):
```
{{ config(
    location = 's3://my-bucket/data/out/' ~ this.name
) }}
```

Reply:
Latest feedback on the slack thread seem to confirm we can accomplish this with something like the following:

```
{{ config(location='/mnt/data/tables/' ~ this.schema ~ '/' ~ this.table) }}
```

...where `~` is the concatenation operator and `this` is defined as described here: https://docs.getdbt.com/docs/this

Reply:
Thanks, @jtcohen6 . Follow-up question, would this also work if declared at the project-level or schema-level? In other words, can it be "late evaluated" to be configured at the module level but then evaluated separately for each table?

Reply:
No, unfortunately. The `this` variable isn't available at the time that `dbt_project.yml` is being parsed. I'd prefer not to have separate config options for `location` and `location_root`. What you _could_ do is:
* Use a macro to define the model's location as a function of its name, or a var to set the root location, and call the macro or var in each model's `location` config.
* Create a config-specific function that merges (concatenates) the `location` set in `dbt_project.yml` with the `location` set in the model `config` block. @drewbanin This feels related to https://github.com/fishtown-analytics/dbt/pull/2062, and it might be more trouble than it's worth.

Reply:
> * Use a macro to define the model's location as a function of its name, or a var to set the root location, and call the macro or var in each model's `location` config.

@jtcohen6 - I'm interested in what this option might look like but I'm not sure I fully understand - do you mean to create a macro (basically a default calculation) that determines location as a function of model name? I think we still end up having two variables - one for `location_root` (which can be set at a project level) and one for `location` (which is overridable per table). Or would we use the same variable name but the behavior would be slightly adapted depending on whether set at the project or table level?

For background, my own planned use case for this would be for disposable spark clusters, where we may want all or most of the ELT outputs (models) to be saved directly to durable locations (such as s3). In this case, we would want an entire schema to default to a specific root directory without having to manually configure a location for each model. 

Do I understand correctly with the option quoted above that we still would need two config variables solving if `this` isn't accessible for project-level config? Can you say a little more about this?

Thanks!

Reply:
I just chatted with @drewbanin, and we agreed on an approach of _always_ appending the model name/alias to the custom `location`, if a `location` is specified. That way, we avoid any collision between tables of different column schemata in the same file path, which Databricks warns about [here](https://docs.databricks.com/spark/latest/spark-sql/language-manual/create-table.html). We would encode that concatenation logic in the `location_clause()` macro.

In the medium term, we hope that we could make this work by allowing you to pass a "late-evaluating" Jinja expression in the `dbt_project.yml` config. Until that time, this feels like a sensible solution. It's hard to see why you _wouldn't_ want the model name included in the file path.

Reply:
Awesome. 👍 For my part, this sounds like a good solution!

Reply:
@jtcohen6 maybe it's better to rename `location` to `location_root`, to make it explicit that we're going to add the model name/alias. Eg:
```
{%- set location_root = config.get('location_root', validator=validation.any[basestring]) -%}
  {%- if location_root is not none %}
    location '{{ location_root }}/{{ this.name }}'
....
```

Reply:
I'm amenable to that naming

Reply:
@NielsZeilemaker Let's go ahead with the `location_root` naming. I think the code should look like:
```
{%- set location_root = config.get('location_root', validator=validation.any[basestring]) -%}
{%- set identifier = model['alias'] -%}
  {%- if location_root is not none %}
    location '{{ location_root }}/{{ identifier }}'
....
```